components/autocomplete/index.js

```diff
@@ -225,22 +226,85 @@ export class Autocomplete extends Component {
 		this.node = node;
 	}
 
-	insertCompletion( range, replacement ) {
-		const container = document.createElement( 'div' );
-		container.innerHTML = renderToString( replacement );
-		while ( container.firstChild ) {
-			const child = container.firstChild;
-			container.removeChild( child );
-			range.insertNode( child );
-			range.setStartAfter( child );
+	insertCompletion( range, replacement, completerName ) {
+		// Wrap completions so we can treat them as tokens with editing boundaries.
+		const tokenWrapper = document.createElement( 'span' );
+		tokenWrapper.innerHTML = renderToString( replacement );
+
+		// Remember the completer in case the user wants to edit the token.
+		tokenWrapper.dataset.autocompleter = completerName;
+
+		// Add classes for general and completer-specific styling.
+		tokenWrapper.classList.add( 'autocomplete-token' );
+		tokenWrapper.classList.add( `autocomplete-token-${ completerName }` );
+
+		const existingTokenWrapper = this.getTokenWrapperNode( range.startContainer );
+		if ( existingTokenWrapper ) {
+			/**
+			 * If we're within an existing token, we want to replace it rather
+			 * than inserting a completion within a completion.
+			 */
+			existingTokenWrapper.parentNode.replaceChild( tokenWrapper, existingTokenWrapper );
```

Comment:
I think all these DOM manipulations need to we wrapped in [`editor.undoManager.transact`](https://www.tinymce.com/docs/api/tinymce/tinymce.undomanager/#transact) so that we are sure this is one history record. Becomes harder to do though as this component is not aware of the TinyMCE editor instance.